src/main/scala/akka/stream/contrib/ZipToStreamFlow.scala

```diff
@@ -0,0 +1,187 @@
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with StageLogging {
+      private val buffer = new ZipBuffer(bufferSize)
+      private var currentStream: Option[InputStream] = None
+      private var emptyStream = true
+
+      setHandler(
+        out,
+        new OutHandler {
+          override def onPull(): Unit =
+            if (isClosed(in)) {
+              if (buffer.isEmpty) {
+                completeStage()
+              } else {
+                buffer.close
+                push(out, buffer.toByteString)
+              }
+            } else pull(in)
+
+          override def onDownstreamFinish(): Unit = {
+            closeInput()
+            buffer.close
+            super.onDownstreamFinish()
+          }
+        }
+      )
+
+      setHandler(
+        in,
+        new InHandler {
+          override def onPush(): Unit = {
+            val zipSource = grab(in)
+            emptyStream = false
+            buffer.startEntry(zipSource.filePath)
+            val stream = zipSource.streamGenerator()
+            currentStream = Some(stream)
+            emitMultiple(out, fileChunks(stream, buffer), () => {
+              buffer.endEntry()
+              closeInput()
+            })
+          }
+
+          override def onUpstreamFinish(): Unit =
+            if (!buffer.isEmpty) {
+              buffer.close()
+              if (isAvailable(out)) {
+                push(out, buffer.toByteString)
+              }
+            } else if (emptyStream) super.onUpstreamFinish()
+        }
+      )
+
+      private def closeInput(): Unit = {
+        currentStream.foreach(_.close)
+        currentStream = None
+      }
+
+      private def fileChunks(stream: InputStream, buffer: ZipBuffer): Iterator[ByteString] = {
+        // This seems like a good trade-off between single-byte
+        // read I/O performance and doubling the ZipBuffer size.
+        //
+        // And it's still a decent defense against DDOS resource
+        // limit attacks.
+        val readBuffer = new Array[Byte](1024)
+        var done = false
+
+        def result: Stream[ByteString] =
+          if (done) Stream.empty
+          else {
+            try {
+              while (!done && buffer.remaining > 0) {
+                val bytesToRead = Math.min(readBuffer.length, buffer.remaining)
+                val count = stream.read(readBuffer, 0, bytesToRead)
+                if (count == -1) {
+                  stream.close
+                  done = true
+                } else buffer.write(readBuffer, count)
+              }
+              buffer.toByteString #:: result
+            } catch {
+              case NonFatal(e) =>
+                closeInput()
+                throw e
+            }
+          }
+
+        result.iterator.filter(_.nonEmpty)
+      }
+    }
+}
+
+object ZipToStreamFlow {
+  final case class ZipSource(filePath: String, streamGenerator: () => InputStream)
```

Comment:
It might be nicer to use Akka streams in this API (even if they'd eventually need to be converted to `InputStream` - WDYT?

Reply:
Yes. When I wrote this, our need was for zipping multiple S3-contained assets to the browser, and the AWS API (at least that we were using) provided synchronous access to the InputStreams. There was no Alpakka, and Akka was just joining our tech stack (I still have the original Play Iteratee version as a gist).

I see this as a good start for an often-needed utility that can be greatly improved.

Reply:
I also prefer the implementation to use Akka Streams API. Convertion from Java stream to Akka Stream is always possible with `StreamConverters`: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala#L25